### PR TITLE
Add realtime container execution logs for BatchOperator

### LINKED ISSUE #31675: Batch Operator: Realtime container execution logs in Airflow task execution log
### Description

Add functionality to Airflow BatchOperator to capture the task execution logs as they happen from AWS CloudWatch to the BatchOperator's task execution logs.

### Use case/motivation

When executing Batch tasks via BatchOperator the logs are not readily available in Airflow task during execution; only a link is provided to point to where the logs can be viewed in CloudWatch. As a result one should navigate the task logs in Airflow to get a handle of the container log path in AWS Cloudwatch and then access the log from AWS Cloudwatch in realtime.

The current behavior introduces the following challenge:

The ongoing execution logs are not captured in Airflow task ( BatchOperator ) log. This typically involves couple of additional steps - which can be avoided.

### Related issues

This is almost identical to the ECSOperator realtime log streaming that was raised in this issue: https://github.com/apache/airflow/issues/22512. In fact most of the description, use case, and motivation are lifted from that issue.

An `EcsTaskLogFetcher` class already exists in `airflow.providers.amazon.aws.hooks.ecs` which AWS Batch can use as a backend. It just needs to be initialized, started, and stopped at the correct time. In the event of an array job, the Batch Operator already defaults to the first task and the same could be done when choosing which log stream to fetch.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by killua1zoldyck**
closes: #31675 

A new utility called `AWSTaskLogFetcher` has been implemented to address the common functionality shared by `EcsTaskLogFetcher` in `EcsRunTaskOperator` and for the `BatchOperator` for this issue.

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/batch_waiters.py [hunk 1]
```diff
@@ -194,6 +201,9 @@ def wait_for_job(self, job_id: str, delay: int | float | None = None) -> None:
 
         :param delay:  A delay before polling for job status
 
+        :param get_batch_log_fetcher: A method that returns batch_log_fetcher of
+            type AwsTaskLogFetcher or None.
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/hooks/batch_waiters.py]: This should explain what returning None from the getter means.

Comment (killua1zoldyck) [reply airflow/providers/amazon/aws/hooks/batch_waiters.py]: I have updated it. Thanks!

**Comment by o-nikolas** [comment]
Note that we have some other changes in flight right now touching the same code (https://github.com/apache/airflow/pull/31824/) CC @vandonr-amz

**Comment by o-nikolas** [comment]
> I did not make any changes to the TaskLogFetcher logic. Since I had to reuse the same LogFetcher without any change I just moved it to `utils/task_log_fetcher.py`. Shall I wait for #31824 to get approved and merge that with my changes and update this PR?

I think it's fine for everyone to proceed as normal and the second person to merge will have to rebase. I just wanted everyone to be aware of each other's changes. The behaviour of the log fetcher is changing.

**Comment by killua1zoldyck** [comment]
Ok. I will wait for any further comments/changes.

**Comment by killua1zoldyck** [comment]
I have updated the changes from #31824 @o-nikolas

**Review by vandonr-amz — APPROVED** [review]

**Comment by o-nikolas** [comment]
> I have updated the changes from #31824 @o-nikolas

Thanks! Unfortunately there are some more conflicts due to some more logging PRs being merged

**Comment by killua1zoldyck** [comment]
> > I have updated the changes from #31824 @o-nikolas
> 
> Thanks! Unfortunately there are some more conflicts due to some more logging PRs being merged

I have updated the PR

**Review by vandonr-amz — APPROVED** [review]
LGTM

**Review by o-nikolas — APPROVED** [review]

**Pull request merged**